kernel/src/net/socket/ip/stream/init.rs

```diff
@@ -63,13 +80,44 @@ impl InitStream {
         option: &RawTcpOption,
         observer: StreamObserver,
     ) -> core::result::Result<ConnectingStream, (Error, Self)> {
-        let bound_port = match self {
-            InitStream::Bound(bound_port) => bound_port,
-            InitStream::Unbound => self.bind_to_ephemeral_endpoint(remote_endpoint)?,
+        // `check_connect` should be called before calling `connect`.
+        debug_assert!(self.can_connect_new);
+
+        let bound_port = if let Some(bound_port) = self.bound_port {
+            bound_port
+        } else {
+            let endpoint = get_ephemeral_endpoint(remote_endpoint);
+            match bind_port(&endpoint, false) {
+                Ok(bound_port) => bound_port,
+                Err(err) => return Err((err, self)),
+            }
         };
 
-        ConnectingStream::new(bound_port, *remote_endpoint, option, observer)
-            .map_err(|(err, bound_port)| (err, InitStream::Bound(bound_port)))
+        ConnectingStream::new(bound_port, *remote_endpoint, option, observer).map_err(
+            |(err, bound_port)| {
+                if err.error() == Errno::ECONNREFUSED {
+                    (err, InitStream::new_refused(bound_port))
+                } else {
+                    (err, InitStream::new_bound(bound_port))
+                }
+            },
+        )
+    }
+
+    pub fn check_connect(&mut self) -> Result<()> {
```

Comment:
Maybe better renamed as `check_and_reset_confused` or `check_and_clear_confused`?

Reply:
Now it's `finish_last_connect`. Please see the latest code.